References
==========

.. [i] Cibulskis K et al. Sensitive detection of somatic point mutations in impure
   and heterogeneous cancer samples. Nat Biotechnol 2013, 31:213-219.

.. [ii] Danecek P, Auton A, Abecasis G, Albers CA, Banks E, DePristo MA, et al. The
   variant call format and VCFtools. Bioinformatics 2011; 27: 2156–8.

.. [iii] Kim SY, Speed TP. Comparing somatic mutation-callers: beyond Venn diagrams.
   BMC Bioinformatics 2013, 14:189.

.. [iv] Koboldt DC et al. VarScan 2: somatic mutation and copy number alteration
   discovery in cancer by exome sequencing. Genome Res 2012, 22:568-576.

.. [v] O'Rawe J et al. Low concordance of multiple variant-calling pipelines:
   practical implications for exome and genome sequencing. Genome Med 2013, 5:28.

.. [vi] Saunders CT et al. Strelka: accurate somatic small-variant calling from
   sequenced tumor-normal sample pairs. Bioinformatics 2012, 28:1811–7.

.. [vii] Xu H. et al. Comparison of somatic mutation calling methods in amplicon and
   whole exome sequence data. BMC Genomics 2014, 15: 244. 
